### DISCUSSION TITLE
Solution to cross-currency payment feature

### DISCUSSION THREAD

**Discussion by trinitys7** [akash-network/community#309] [Ideas]
# AKT fee abstraction

Hi Akash Team. I'm Nguyen from Caelus Labs.  We have idea about [Stable payments & Take rates]. It's related with `Phase 2 - Fast-follow` and `Phase 3 - Bells and whistles` in this discusstion : https://github.com/orgs/akash-network/discussions/147
(https://github.com/orgs/akash-network/discussions/147).  We called it `AKT Fee abstraction`

## Context
Currently, Akash Network only allows the use of AKT tokens for its services. Users who hold tokens other than AKT need to swap their tokens for AKT if they want to use Akash services.

With `AKT fee abstraction` implementation, user can use IBC token for service fee in Akash Network. Module will swap IBC token via Osmosis automatically and dont't break current payment flow. 

The usecases that motivated the development of this module are:
 - More flexiable payment : this module allow the use of IBC token to pay service fees in Akash Network
 - Additional payment options for the Bid/Lease mechanism in Akash Network, with `AKT fee abstraction`. 
     - Providers able to choose the types of tokens they accept for their services, as currently they only receive AKT. 
     - Users who want to lease a service can use any type of token as a service fee.

Otherwise, using fee abstraction promote the AKT trading volumn in DEX.

## Prototype

Akash fee abstraction mechanism :

 1. Get price data from osmosis (this is on-chain data) :
- Periodically pulling `twap data` from osmosis by ibc-ing to `async-icq` module on Osmosis, this `twap data` will update the exchange rate of osmosis to customer chain's native token.
- This exchange rate is used to calaculate swap out amount.

 2. Handling bid/lease with ibc-token fee:
- The core change is in `escrow` and `market` module.
- For bid/lease process :
    - Providers can choose IBC token for their payment when they `CreateBid`. When a user `CreateLease`, the module calculates and converts IBC tokens to the needed amount for the bid. In this process we will use osmosis's Cross chain Swap (XCS) feature for converting token.
    - User can deposit IBC token into escrow address (by `DepositDeployment` msg), this amount will be automatically convert to token with denom specified when provider `CreateBid`.

####  1. Get price data from osmosis (this is on-chain data) 
The exchange rate will be set by TWAP data from osmosis DEX via `async-icq`.  

This exchange rate will be used to calculate `EstimatedLeaseAmount` when user leasing by`CreateLease` msg.

#### 2.Allow providers to use IBC token in their bid when `CreateBid`.

Providers can choose a token for payment when they CreateBid. When a user creates a lease by `CreateLease` msg, the fee abstraction module calculates the IBC token needed by `twap data` : 

    `EstimatedLeaseAmount = bidAmount * exchangeTwapRate`

The module will convert the `EstimatedLeaseAmount` to the required bid by swapping this token via XCS. Once this is completed, we will have the `FractionalPayment` with the bid token and process it the same way as the current process.

#### 3. Allow user deposit IBC token when `DepositDeployment`.

If a user wants to `DepositDeployment` with IBC tokens, the module will first check if that token is allowed (allowed denom is a module parameter). Then, it will use Osmosis's Cross Chain Swap (XCS) feature to convert the IBC token to the required bid token (specified by the provider when `CreateBid`) via the Osmosis DEX.

Essentially, the module transfers IBC tokens to the Osmosis Cross Chain Swap contract with a custom memo to swap the IBC token for the matched bid token, and then transfers them back to the escrow address in Akash Network.

**Comment by troian** [comment]
Hi @trinitys7 thank you for opening up discussion.

The headline and description are rather confusing.
Isn't `AKT Fee Abstraction` meant to allow various tokens to be used for gas payment.

A few remarks tho:
1. What happens if IBC relayers are down for an hour/day?
2. This is not how bid/lease works. The provider cannot choose what token to use for payment settlement. It is done by the tenant during deployment creation.
3. Using non-native tokens for deployment has already been implemented and has been on the mainnet since v0.24.0 upgrade

edit: tho there is `async-icq` implementation, it has to go out of the draft stage, and be under cosmos-sdk team before considering it's adoption by the akash-network.

**Reply by trinitys7** [reply]
thanks for your feedback, we are going to close this discussion here.

**Comment by trinitys7** [comment]
As per the sig chain call, it is decided that the cross-currency payment feature is not needed, instead a more useful use case for osmosis dex integration would be to support USDC holder paying gas with USDC instead of AKT (which could be realized by something similar to the fee abs module) . Therefore, we'll close this discussion and open a new one to talk about ideas regard to the USDC gas payment feature.
